onnxruntime/core/providers/webnn/builders/impl/lrn_op_builder.cc

```diff
@@ -0,0 +1,149 @@
+  const uint32_t trailing_padding = ceil((size - 1) / 2);
+  std::vector<uint32_t> beginning_padding{0, 0, 0, leading_padding};
+  std::vector<uint32_t> ending_padding{0, 0, 0, trailing_padding};
+  emscripten::val pad_options = emscripten::val::object();
+  pad_options.set("label", node_name + "_pad");
+  emscripten::val pad_output =
+      wnn_builder.call<emscripten::val>("pad", pow1_output, emscripten::val::array(beginning_padding),
+                                        emscripten::val::array(ending_padding), pad_options);
+
+  // averagePool2d(pad_output, pool_potions)
+  const std::vector<uint32_t> kernel_shape{1, size};
+  emscripten::val pool_options = emscripten::val::object();
+  pool_options.set("label", node_name + "_averagePool2d");
+  pool_options.set("windowDimensions", emscripten::val::array(kernel_shape));
+  emscripten::val pool_output = wnn_builder.call<emscripten::val>("averagePool2d", pad_output, pool_options);
+
+  // transpose(pool_output, permutation=[0, 3, 1, 2])
+  // Move dimension 3 back to dimension 1.
+  if (model_builder.GetPreferredLayout() == DataLayout::NCHW) {
+    std::vector<uint32_t> perm{0, 3, 1, 2};
+    emscripten::val transpose_options = emscripten::val::object();
+    transpose_options.set("label", node_name + "_transpose_inverse");
+    transpose_options.set("permutation", emscripten::val::array(perm));
+    pool_output =
+        wnn_builder.call<emscripten::val>("transpose", pool_output, transpose_options);
+  }
+
+  // mul(pool_output, alpha_contant)
+  label_options.set("label", node_name + "_mul");
+  emscripten::val mul_output =
+      wnn_builder.call<emscripten::val>("mul", pool_output, alpha_constant, label_options);
+
+  // add(mul_output, bias_contant)
+  label_options.set("label", node_name + "_add");
+  emscripten::val add_output = wnn_builder.call<emscripten::val>("add", mul_output, bias_constant, label_options);
+
+  // pow(add_output, beta_constant)
+  label_options.set("label", node_name + "_pow2");
+  emscripten::val pow2_output = wnn_builder.call<emscripten::val>("pow", add_output, beta_constant, label_options);
+
+  // div(input, pow2_output)
+  label_options.set("label", node_name + "_div");
+  emscripten::val div_output = wnn_builder.call<emscripten::val>("div", input, pow2_output, label_options);
+
+  model_builder.AddOperand(node.OutputDefs()[0]->Name(), std::move(div_output));
+  return Status::OK();
+}
+
+// Operator support related.
+bool LRNOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers,
+                                     const Node& node,
+                                     const WebnnDeviceType /* device_type */,
+                                     const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger))
+    return false;
+  const auto input_size = input_shape.size();
+  if (input_size != 4) {
```

Comment:
LRN supports 2D up to ND. Is it tough to extend it, if `beginning_padding` and `ending_padding` were extended to dynamic size? It's okay if it's not worth it, as 4D is the most common case.

Reply:
It should not be difficult to expand `beginning_padding` and `ending_padding` to dynamic sizes, but the following averagePool2d operator only supports 4D tensors as input. We should be able to implement ND LRN by inserting reshape operators before and after the averagePool2d operator, but this undoubtedly increases the complexity of decomposition. We can consider expanding it later when we encounter ND LRN in the real model. WDYT?

Reply:
> We should be able to implement ND LRN by inserting reshape operators before and after the averagePool2d operator,
> but the following averagePool2d operator only supports 4D tensors as input

🤔 Indeed, that's annoying. I think we should relax WebNN's requirement on the pooling ops so they don't require 4D (2D and 3D should be allowed too imo), so the caller doesn't have to add such reshapes around the ops. I also proposed to @huningxin on an email thread with Austin that we pass explicit axes (like reduction) to the pooling ops.

Anyway, we can revisit it later, but 4D is fine for now.

Reply:
> I also proposed to @huningxin on an email thread with Austin that we pass explicit axes (like reduction) to the pooling ops.

This is a good use case for ND pooling. We should put that in the spec issue.